tests/python/gaia-ui-tests/gaiatest/tests/functional/settings/test_settings_bluetooth.py

```diff
@@ -34,6 +32,10 @@ def test_toggle_bluetooth_settings(self):
         self.assertEquals(bluetooth_settings.device_name, device_name)
 
         bluetooth_settings.enable_visible_to_all()
-
         self.assertTrue(self.data_layer.bluetooth_is_discoverable)
         self.assertEquals(self.data_layer.bluetooth_name, device_name)
+
+    def tearDown(self):
+        # Disable bluetooth to reset the state
+        from gaiatest.apps.settings.regions.bluetooth import Bluetooth
+        Bluetooth(self.marionette).disable_bluetooth()
```

Comment:
I think you need to to do, instead of what you're doing here:
self.marionette.switch_to_frame()
self.data_layer.bluetooth_disable()
GaiaTestCase.tearDown(self)

You can't rely on UI in tearDown, because the test can fail in various places, where the settings app might not be even there.
So you need to reset it programmatically.